build: add LTO as a configure option

Starting Fedora-33, LTO will be enabled by default for package builds. 
Therefore when creating glusterfs RPMs [using `make glusterrpms`] the following options are added automatically to the build: `-flto=auto -ffat-lto-objects`  [tested on Fedora:33 image].

Benefits of building with LTO:
--  smaller, faster executables, remove dead code, like uncalled non-static functions [2] (glusterfs binary size reduced by~3%)
-- LTO allows deeper analysis of package source code at compile time which can improve various GCC diagnostics and thus improve our ability to catch bugs at compile time [1]

WIP:
Comparing the result of [gluster-performance-test-suite](https://github.com/gluster/gluster-performance-test-suite) between LTO-build against none-LTO didn't show any improvement. (setup was local VMs), I plan to rerun it on physical machines.

@msaju, @kalebskeithley, @mykaul 
do you think it's worth adding LTO as a configure option to the build when cc=gcc-10? if so should it be enabled by default? 
looks that problems such as compile-time and memory usage were fixed

[1] https://fedoraproject.org/wiki/LTOByDefault#Changes.2FLTO_by_default_for_package_builds
[2] https://lwn.net/Articles/744507/

Warnings list with gcc-10 and LTO, the `-Wstringop-overflow=` warning addressed here: https://github.com/gluster/glusterfs/pull/1747
The `-Wstringop-truncation` warning suppressed when building with LTO
```
# grep -E warning.*=  F33_source_make.LTO
afr-common.c:7512:19: warning: 'strncat' specified bound 36 equals source length [-Wstringop-overflow=]
afr-common.c:7527:19: warning: 'strncat' specified bound depends on the length of the source argument [-Wstringop-overflow=]
changelog-helpers.c:1971:9: warning: 'strncat' specified bound depends on the length of the source argument [-Wstringop-overflow=]
glusterd-utils.c:14919:9: warning: 'strncpy' specified bound depends on the length of the source argument [-Wstringop-overflow=]
glusterd-utils.c:15065:27: warning: 'strncat' specified bound depends on the length of the source argument [-Wstringop-overflow=]
glusterd-utils.c:15074:35: warning: 'strncat' specified bound depends on the length of the source argument [-Wstringop-overflow=]
glusterd-volgen.c:3759:9: warning: 'strncat' specified bound depends on the length of the source argument [-Wstringop-overflow=]
glusterd-volgen.c:3780:21: warning: 'strncat' specified bound depends on the length of the source argument [-Wstringop-overflow=]
```

Warnings list with gcc-10 without LTO:
```
[root@tshacked hostdata]# grep -E warning.*=  F32_source_make.NO_LTO 
graph.c:494:33: warning: ‘%s’ directive output may be truncated writing up to 255 bytes into a region of size 125 [-Wformat-truncation=]
afr-common.c:7527:19: warning: ‘strncat’ specified bound depends on the length of the source argument [-Wstringop-overflow=]
afr-common.c:7512:19: warning: ‘strncat’ specified bound depends on the length of the source argument [-Wstringop-overflow=]
posix-inode-fd-ops.c:3436:47: warning: ‘%s’ directive output may be truncated writing up to 4081 bytes into a region of size 37 [-Wformat-truncation=]
changelog-helpers.c:1971:9: warning: ‘strncat’ specified bound depends on the length of the source argument [-Wstringop-overflow=]
glusterd-utils.c:14919:9: warning: ‘strncpy’ specified bound depends on the length of the source argument [-Wstringop-overflow=]
glusterd-utils.c:15065:27: warning: ‘strncat’ specified bound depends on the length of the source argument [-Wstringop-overflow=]
glusterd-utils.c:15074:35: warning: ‘strncat’ specified bound depends on the length of the source argument [-Wstringop-overflow=]
glusterd-store.c:4376:36: warning: ‘snprintf’ output may be truncated before the last format character [-Wformat-truncation=]
glusterd-store.c:4362:36: warning: ‘snprintf’ output may be truncated before the last format character [-Wformat-truncation=]
glusterd-volgen.c:3759:9: warning: ‘strncat’ specified bound depends on the length of the source argument [-Wstringop-overflow=]
glusterd-volgen.c:3780:21: warning: ‘strncat’ specified bound depends on the length of the source argument [-Wstringop-overflow=]
```

--- a/configure.ac
+++ b/configure.ac
@@ -1351,11 +1351,22 @@ case $host_os in
         ;;
 esac
 
-CC_VER=`"$CC" -dumpversion 2>/dev/null`
-AS_IF([ test ! -z "$CC_VER" && test "$CC_VER" -ge "10" ], [
-  	GF_CFLAGS="${GF_CFLAGS} -flto"
-	GF_LDFLAGS="${GF_LDFLAGS} -flto"
-	AC_MSG_NOTICE([building with LTO])
+# LTO section
+AC_ARG_ENABLE([lto],
+              AC_HELP_STRING([--disable-lto],
+                             [Disable buidling with LTO]))
+
+LTO_BUILD=no
+AS_IF([ test "x$enable_lto" != "xno" ], [
+    CC_VER=`"$CC" -dumpversion 2>/dev/null`
+    AS_IF([ test ! -z "$CC_VER" && test "$CC_VER" -ge "10" ], [
+        GF_CFLAGS="${GF_CFLAGS} -flto"
+        GF_LDFLAGS="${GF_LDFLAGS} -flto"
+        LTO_BUILD=yes
+        AC_MSG_NOTICE([building with LTO])],
+        [AC_MSG_NOTICE([not using LTO for gcc_ver=$CC_VER]) ]) 
+    ], [
+    AC_MSG_NOTICE([LTO is disabled])
 ])
 
 # Default value for sbindir
@@ -1735,6 +1746,7 @@ echo "Cloudsync            : $BUILD_CLOUDSYNC"
 echo "Metadata dispersal   : $BUILD_METADISP"
 echo "Link with TCMALLOC   : $BUILD_TCMALLOC"
 echo "Enable Brick Mux     : $USE_BRICKMUX"
+echo "Building with LTO    : $LTO_BUILD"
 echo
 
 # dnl Note: ${X^^} capitalization assumes bash >= 4.x

--- a/configure.ac
+++ b/configure.ac
@@ -1351,6 +1351,13 @@ case $host_os in
         ;;
 esac
 
+CC_VER=`"$CC" -dumpversion 2>/dev/null`
+AS_IF([ test ! -z "$CC_VER" && test "$CC_VER" -ge "10" ], [
+  	GF_CFLAGS="${GF_CFLAGS} -flto"
+	GF_LDFLAGS="${GF_LDFLAGS} -flto"
+	AC_MSG_NOTICE([building with LTO])
+])
+
 # Default value for sbindir
 prefix_temp=$prefix
 exec_prefix_temp=$exec_prefix

